Test: DNS Server in test networks

omerfirmak opened the pull request:
Fixes #2789

Geod24 committed:
Fix #2558: Add a client DNS resolver

This will allow Agora to resolves validators public key to IP address and protocols,
bridging the gap between the HTTP and DNS capabilities of the name registry.

Co-Authored-By: Muhammed Kadir Yücel <muhammed.yucel@bosagora.io>

omerfirmak committed:
DNSResolver: Make it an abstract class

So that we can extend it for Vibe and LocalRest separately

omerfirmak committed:
Test: Implement LocalRestDNSResolver

omerfirmak committed:
Instantiate DNSResolver in both tests and regular builds

omerfirmak committed:
Test: Start DNS Server

omerfirmak committed:
Test: DNS testbed

- Register IPv4 addresses for Validators, so that they are compatible
with DNS
- Set unittest realm for network tests
- An example test to query LocalRest DNS

omerfirmak commented:
Added a very simple testbed as well, DNS can now be unittested.

Next step is to integrate DNSResolver into Agora.

omerfirmak commented:
> Next step is to integrate DNSResolver into Agora.

Which will have it's own PR at some point, hopefully soon.

mkykadir commented on source/agora/network/Manager.d:
```diff
@@ -463,6 +464,7 @@ public class NetworkManager
         this.discovery_task = new AddressDiscoveryTask(&this.addAddresses);
         this.clock = clock;
         this.owner_node = owner_node;
+        this.resolver = resolver;
```
Ok, this is a decision point and may be conflicting with #2703 's caching registry.
I think (we might also rename it but) we should decide tasks of the `DNSResolver`. Currently it asks queries to configured peers and needs an implementation to cache and manage TTLs (which the mentioned PR implements it on `Registry` since we already have a storage there). After the mentioned PR we can have all nodes to have registry with type `caching` enabled by default, thus requesting a `getValidator` will eventually pass through its own registry and manager will not need this.

omerfirmak replied:
Well, we could just add the local caching registry to `resolv.conf` and utilize it that way.
NetworkClient will have to handle expiration of TTL. But I beleive those are not within the scope of this PR

mkykadir replied:
Yeah, #2703 already implements a caching registry with TTL support.
^ With that, then this PR also shouldn't include a resolver in the Manager, might be too early for mentioned reasons.

omerfirmak committed:
Manager: remove getNetworkClient

It was a wrapper with no real use, not overriden by tests.
Tests override getClient.

omerfirmak committed:
DNSResolver: Acquire new UDP port per request

Otherwise we risk multiple Fibers listening on the same socket
and out-of-order responses

omerfirmak committed:
DNSResolver: Lookup for an A record by default

DNS servers in the wild are not guaranteed to implement QTYPE.ALL

omerfirmak committed:
DNSResolver: Fix CNAME recursion

We were looking up same name over and over.

omerfirmak committed:
Manager&Client: Integrate DNSResolver

NetworkClient will now use DNSResolver for hostname lookups with
no regard to TTL just yet.

omerfirmak commented:
Updated, a FullNode running on my local machine can connect to test network with no problems.

mkykadir commented on source/agora/network/DNSResolver.d:
```diff
@@ -0,0 +1,283 @@
+    public abstract ResourceRecord[] query (const(char)[] name, QTYPE type = QTYPE.ALL) @trusted;
+
+    ///
+    public Message buildQuery (const(char)[] name, QTYPE type = QTYPE.ALL) @safe
```
#2872 marks it as a `protected`, which should be `public` I believe

omerfirmak commented:
Commits of this PR was mostly  merged in #2872 and the remaining are most likely will never make in. So closing this one
